### ISSUE TITLE
[prefer-function-type] false positive on global augmentation

### ISSUE THREAD

**Issue by mysticatea** [typescript-eslint/typescript-eslint#454]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-function-type": "error"
  }
}
```

```TS
export {}
declare global {
    interface Function {
        (...args: any[]): any
    }
}
```

**Expected Result**

No errors because the type alias with a function type causes a syntax error (`Duplicate identifier 'Function'`).

**Actual Result**

```
  5:9  error  Interface has only a call signature - use 'type Function = (...args: any[]) => any' instead  @typescript-eslint/prefer-function-type
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.7.0` |
| `@typescript-eslint/parser`        | `1.7.0` |
| `TypeScript`                       | `3.4.3` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.9.0` |

**Comment by mysticatea** [comment]
Also, on partial interfaces:

```ts
// This exists in another file.
interface A {
    a: number
}

// This is in this file.
interface A {
    (...args: any[]): any
}
```

**Comment by bradzacher** [comment]
Having a quick look, this requires type information to implement because we can't easily detect augmentation without it.
Will punt on it for now because it's technically breaking to add type information to the rule.

**Comment by mysticatea** [comment]
If it cannot detect the situation, we need to remove autofix from the rule because it generates syntax errors.

**Comment by fasttime** [comment]
I have another false positive like the ones above, but without global augmentation or partial interfaces.

```ts
interface Test { (): any; }

class Test
{
    constructor()
    {
        return (): null => null;
    }
}
```

Where the first line incorrectly autofixes to

```ts
type Test = () => any;
```

**Comment by JoshuaKGoldberg** [comment]
From #6104: we're going to go close this issue out as wontfix. Adding type information to a rule makes the rule slower and restricts the range of developers who can use it.

https://github.com/typescript-eslint/typescript-eslint/pull/6104#pullrequestreview-1328164311 explains why this is a rare enough edge case we don't believe the type info is worth it:

> I'm not going to lie - looking at it - I don't know if we should introduce type info for this case. The cases are so, so, so rare:
> 
> 1. For the first case - not only do you need to be doing global module augmentation, but you also need to be defining a call signature on an interface type within a global augmentation.
> 2. For the second case - you need to be declaration merging a class and an interface (which we already know is honestly dumb and SUPER unsafe - the vast majority of codebases DON'T do this), but you need to be declaration merging a class and an interface where the interface has a call signature.
> 
> The cases are so rare that forcing every user to use type info for the case just doesn't seem like it's good for the ecosystem.
> 
> As a data point for how rare this is - since that issue was file way back in 2019 it has received engagement from just 1 additional community member. It's just not a problem we run into.
> 
> Instead of adding type-info, I propose one of two approaches:
> 
> 1. specifically add handling for these cases and _disable_ the fixer, whilst still reporting. A false positive is fine - that be disabled and ignored, but fixing incorrectly is not. Disabling the fixer allows this case
> 2. document this rare edge case in the rule doc and move on.
